sdk/search/Azure.Search/src/Models/IndexDocumentsAction{T}.cs

```diff
@@ -0,0 +1,216 @@
+            using JsonDocument nested = JsonDocument.Parse(json);
+            if (nested.RootElement.ValueKind == JsonValueKind.Object)
+            {
+                foreach (JsonProperty property in nested.RootElement.EnumerateObject())
+                {
+                    property.WriteTo(writer);
+                }
+            }
+
+            writer.WriteEndObject();
+        }
+    }
+
+    /// <summary>
+    /// Creates <see cref="IndexDocumentsAction{T}"/> instances to add to an
+    /// <see cref="IndexDocumentsBatch{T}"/> and submit via
+    /// <see cref="SearchIndexClient.IndexDocuments"/>.
+    /// </summary>
+    public static class IndexDocumentsAction
+    {
+        /// <summary>
+        /// Create an <see cref="IndexDocumentsAction{T}"/> to upload.
+        /// </summary>
+        /// <typeparam name="T">
+        /// The .NET type that maps to the index schema. Instances of this type
+        /// can be retrieved as documents from the index.
+        /// </typeparam>
+        /// <param name="document">The document to upload.</param>
+        /// <returns>
+        /// An <see cref="IndexDocumentsAction{T}"/> to upload.
+        /// </returns>
+        public static IndexDocumentsAction<T> Upload<T>(T document) =>
+            new IndexDocumentsAction<T>(IndexActionType.Upload, document);
+
+        /// <summary>
+        /// Create an <see cref="IndexDocumentsAction{SearchDocument}"/> to
+        /// upload.
+        /// </summary>
+        /// <param name="document">The document to upload.</param>
+        /// <returns>
+        /// An <see cref="IndexDocumentsAction{SearchDocument}"/> to upload.
+        /// </returns>
+        public static IndexDocumentsAction<SearchDocument> Upload(SearchDocument document) =>
+            new IndexDocumentsAction<SearchDocument>(IndexActionType.Upload, document);
+
+        /// <summary>
+        /// Create an <see cref="IndexDocumentsAction{T}"/> to merge.
+        /// </summary>
+        /// <typeparam name="T">
+        /// The .NET type that maps to the index schema. Instances of this type
+        /// can be retrieved as documents from the index.
+        /// </typeparam>
+        /// <param name="document">The document to merge.</param>
+        /// <returns>
+        /// An <see cref="IndexDocumentsAction{T}"/> to merge.
+        /// </returns>
+        public static IndexDocumentsAction<T> Merge<T>(T document) =>
+            new IndexDocumentsAction<T>(IndexActionType.Merge, document);
+
+        /// <summary>
+        /// Create an <see cref="IndexDocumentsAction{SearchDocument}"/> to
+        /// merge.
+        /// </summary>
+        /// <param name="document">The document to merge.</param>
+        /// <returns>
+        /// An <see cref="IndexDocumentsAction{SearchDocument}"/> to merge.
+        /// </returns>
+        public static IndexDocumentsAction<SearchDocument> Merge(SearchDocument document) =>
```

Comment:
These overloads don't add actual specializations. What purpose do they serve? Just hints it can be a `SearchDocument` or a custom model?

Reply:
Yes - which mirrors all our other APIs like Search/Search<T>.